File: app/src/main/java/com/shopcart/ui/adapters/SliderAdapter.java
package com.shopcart.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.viewpager.widget.PagerAdapter;

import com.bumptech.glide.Glide;
import com.shopcart.Banner;
import com.shopcart.R;

import java.util.ArrayList;

public class SliderAdapter extends PagerAdapter {

    private final Context context;
    private ArrayList<Banner> list;

    public SliderAdapter(Context context) {
        this.context = context;
    }


    @Override
    public int getCount() {
        if (list != null)
             return list.size();
        return 0;
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {

        LayoutInflater layoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        assert layoutInflater != null;
        View view = layoutInflater.inflate(R.layout.slider_layout, container, false);

        AppCompatImageView banner = view.findViewById(R.id.banner_img_banner);

        if (list != null && list.size() != 0) {
            Glide.with(context).load(list.get(position).getPhotoUrl()).into(banner);
        }

        container.addView(view);
        if (position == getCount() - 1) {
            position = 0;
        }

        return view;
    }

    public void setList(ArrayList<Banner> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((View) object);
    }

}

File: app/src/main/java/com/shopcart/ui/adapters/SliderAdapter.kt
package com.shopcart.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.shopcart.data.models.Banner
import com.shopcart.databinding.SliderLayoutBinding

class SliderAdapter : ListAdapter<Banner, SliderAdapter.ViewHolder>(MyDiffUtil) {

    // DiffUtil is a utility class that calculates the difference between two lists
    // and outputs a list of update operations that converts the first list into the second one.
    companion object MyDiffUtil : DiffUtil.ItemCallback<Banner>() {


        // Checks if the two objects are the same
        override fun areItemsTheSame(oldItem: Banner, newItem: Banner): Boolean {
            return oldItem == newItem
        }

        // Checks if the data between the two objects is the same
        override fun areContentsTheSame(oldItem: Banner, newItem: Banner): Boolean {
            return oldItem.id == newItem.id
        }

    }

    // Holds the views for adding it to image and text
    inner class ViewHolder(val binding: SliderLayoutBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(position: Int) {
            Glide.with(binding.root.context).load(currentList[position].photoUrl)
                .into(binding.sliderImg)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            SliderLayoutBinding.inflate(LayoutInflater.from(parent.context), parent, false)

        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(position)
    }
}